## Annotated dataset

- In this notebook, I will explore the annotated dataset stored as JSON file.
- To annotate text files containing code-switching instances a text annotation feature of [the labelbox platform](https://labelbox.com/product/annotate/text/) was used. 
- First, we will load and experiment with the test sample which contains two objects=annotated files;
- Second, we will load and explore the sample two which contains 14 objects=annotated files;
- Finally, we will carry out a preliminary analysis of annotated data:
   - how many utterances are there? 
   - how many code-switching instances are there within utterance?
   - what type of code-switching (cs) is prevalent: inter-sent, intra-sent or intra-word?
       - Our hypothesis is that intra-sent and intra-word cs types would be prevalent.
   - for each type of cs, which linguistic unit is more common: phrases, discourse markers, particular POS (n,adj,adv, pronoun), etc.

In [142]:
# import labelbox as lb
# import labelbox.types as lb_types
import json
import pandas as pd
from flatten_json import flatten
from collections import defaultdict
import seaborn as sns

## Load annotated files: test sample 1

- Open json files exported from the labelbox annotation project.
    - a structure of json file: nested dictionary with multiple keys:values and values stored as lists;
- Extract data from nested dictionary.
    - tutorial can be found [here](https://towardsdatascience.com/how-do-i-extract-nested-data-in-python-4e7bed37566a) 

In [143]:
with open('/Users/aidyn/Downloads/export-2023-03-16T21_01_57.705Z.json') as f:
   data = json.load(f)

# print(data[:10])

In [144]:
type(data)

list

In [145]:
len(data)
type(data[0])

dict

In [264]:
label_dict = data[0]['Label']
label_dict.keys()

dict_keys(['objects', 'classifications', 'relationships'])

In [147]:
type(data[0]['Label'])

dict

In [148]:
data[0]['Label']['objects'][0]

{'featureId': 'clfbk9dbf0002396jfl1k6gpk',
 'schemaId': 'clfbcpdz013ua07zn9k319jn3',
 'color': '#1CE6FF',
 'title': 'uttr',
 'value': 'uttr',
 'version': 1,
 'format': 'text.location',
 'data': {'location': {'start': 151, 'end': 211}},
 'classifications': [{'featureId': 'clfbk9hrj0004396jaxr9fvb7',
   'schemaId': 'clfbcpdz013ub07zn8cw44wpj',
   'title': 'lang',
   'value': 'lang',
   'position': 0,
   'answer': {'featureId': 'clfbk9hrj0003396jlk95obai',
    'schemaId': 'clfbcpdz013uc07zn8vjl7q45',
    'title': 'kz',
    'value': 'kz',
    'position': 0}}]}

In [149]:
type(data[0]['Label']['objects'])

list

In [150]:
len(data[0]['Label']['objects'])

49

In [151]:
data[0]['Label']['objects'][0]

{'featureId': 'clfbk9dbf0002396jfl1k6gpk',
 'schemaId': 'clfbcpdz013ua07zn9k319jn3',
 'color': '#1CE6FF',
 'title': 'uttr',
 'value': 'uttr',
 'version': 1,
 'format': 'text.location',
 'data': {'location': {'start': 151, 'end': 211}},
 'classifications': [{'featureId': 'clfbk9hrj0004396jaxr9fvb7',
   'schemaId': 'clfbcpdz013ub07zn8cw44wpj',
   'title': 'lang',
   'value': 'lang',
   'position': 0,
   'answer': {'featureId': 'clfbk9hrj0003396jlk95obai',
    'schemaId': 'clfbcpdz013uc07zn8vjl7q45',
    'title': 'kz',
    'value': 'kz',
    'position': 0}}]}

In [152]:
data[0]['Label']['objects'][0]['featureId']

'clfbk9dbf0002396jfl1k6gpk'

In [153]:
data[0]['Label']['objects'][0]['classifications']

[{'featureId': 'clfbk9hrj0004396jaxr9fvb7',
  'schemaId': 'clfbcpdz013ub07zn8cw44wpj',
  'title': 'lang',
  'value': 'lang',
  'position': 0,
  'answer': {'featureId': 'clfbk9hrj0003396jlk95obai',
   'schemaId': 'clfbcpdz013uc07zn8vjl7q45',
   'title': 'kz',
   'value': 'kz',
   'position': 0}}]

In [154]:
type(data[0]['Label']['objects'][0]['classifications'])

list

In [155]:
data[0]['Label']['objects'][0]['classifications'][0]['answer']

{'featureId': 'clfbk9hrj0003396jlk95obai',
 'schemaId': 'clfbcpdz013uc07zn8vjl7q45',
 'title': 'kz',
 'value': 'kz',
 'position': 0}

In [156]:
type(data[0]['Label']['objects'][0]['classifications'][0]['answer'])

dict

In [157]:
for item in range(len(data)):
     print(data[item]['Label']['objects'][0]['classifications'])

[{'featureId': 'clfbk9hrj0004396jaxr9fvb7', 'schemaId': 'clfbcpdz013ub07zn8cw44wpj', 'title': 'lang', 'value': 'lang', 'position': 0, 'answer': {'featureId': 'clfbk9hrj0003396jlk95obai', 'schemaId': 'clfbcpdz013uc07zn8vjl7q45', 'title': 'kz', 'value': 'kz', 'position': 0}}]
[{'featureId': 'clfbkxt8o0004396j79uc1v16', 'schemaId': 'clfbcpdz013ub07zn8cw44wpj', 'title': 'lang', 'value': 'lang', 'position': 0, 'answer': {'featureId': 'clfbkxt8o0003396jgcabgizk', 'schemaId': 'clfbcpdz113ue07zn3mzi80nz', 'title': 'rs', 'value': 'rs', 'position': 1}}]


In [158]:
for item in range(len(data)):
     print(data[item]['Label']['objects'][0])

{'featureId': 'clfbk9dbf0002396jfl1k6gpk', 'schemaId': 'clfbcpdz013ua07zn9k319jn3', 'color': '#1CE6FF', 'title': 'uttr', 'value': 'uttr', 'version': 1, 'format': 'text.location', 'data': {'location': {'start': 151, 'end': 211}}, 'classifications': [{'featureId': 'clfbk9hrj0004396jaxr9fvb7', 'schemaId': 'clfbcpdz013ub07zn8cw44wpj', 'title': 'lang', 'value': 'lang', 'position': 0, 'answer': {'featureId': 'clfbk9hrj0003396jlk95obai', 'schemaId': 'clfbcpdz013uc07zn8vjl7q45', 'title': 'kz', 'value': 'kz', 'position': 0}}]}
{'featureId': 'clfbkxma70002396jwrrwp26n', 'schemaId': 'clfbcpdz013ua07zn9k319jn3', 'color': '#1CE6FF', 'title': 'uttr', 'value': 'uttr', 'version': 1, 'format': 'text.location', 'data': {'location': {'start': 26, 'end': 42}}, 'classifications': [{'featureId': 'clfbkxt8o0004396j79uc1v16', 'schemaId': 'clfbcpdz013ub07zn8cw44wpj', 'title': 'lang', 'value': 'lang', 'position': 0, 'answer': {'featureId': 'clfbkxt8o0003396jgcabgizk', 'schemaId': 'clfbcpdz113ue07zn3mzi80nz', 't

In [159]:
for item in range(len(data)):
     print(data[item]['Label']['objects'][0]['classifications'][0]['answer'])

{'featureId': 'clfbk9hrj0003396jlk95obai', 'schemaId': 'clfbcpdz013uc07zn8vjl7q45', 'title': 'kz', 'value': 'kz', 'position': 0}
{'featureId': 'clfbkxt8o0003396jgcabgizk', 'schemaId': 'clfbcpdz113ue07zn3mzi80nz', 'title': 'rs', 'value': 'rs', 'position': 1}


## Build pd dataframe

- Convert json file to pd df
    - in df keys are column names and objects (with nested dictionaries) are rows;

In [160]:
# reading JSON file
# df = pd.read_json('/Users/aidyn/Downloads/export-2023-03-16T21_01_57.705Z.json')
df = pd.DataFrame(data)
# displaying sample output
df.head()

ID                 DataRow ID  \
0  clfbk8guf04g5071jbrqo26bq  clfbcqidc3xmk07zvgxw373vk   
1  clfbkxgaq0bes07012isv5rq9  clfbkqg468bwf07zc7n31dl3j   

                                        Labeled Data  \
0  https://storage.labelbox.com/clfagslqv28u407zn...   
1  https://storage.labelbox.com/clfagslqv28u407zn...   

                                               Label      Created By  \
0  {'objects': [{'featureId': 'clfbk9dbf0002396jf...  mob75@pitt.edu   
1  {'objects': [{'featureId': 'clfbkxma70002396jw...  mob75@pitt.edu   

       Project Name                Created At                Updated At  \
0  annotation-text1  2023-03-16T20:36:16.000Z  2023-03-16T20:36:16.000Z   
1  annotation-text1  2023-03-16T20:59:51.000Z  2023-03-16T20:59:51.000Z   

   Seconds to Label  Seconds to Review  ...  Agreement Is Benchmark  \
0           691.644             32.957  ...         -1            0   
1           926.531              0.000  ...         -1            0   

  Benchmark Agreement  Benchmark ID    Dataset Name  Reviews  \
0                  -1          None  cs-annotation1       []   
1                  -1          None           text2       []   

                                          View Label Has Open Issues Skipped  \
0  https://editor.labelbox.com?project=clfbk7q941...               0   False   
1  https://editor.labelbox.com?project=clfbk7q941...               0   False   

                               DataRow Workflow Info  
0  {'taskName': 'Done', 'Workflow History': [{'ac...  
1  {'taskName': 'Done', 'Workflow History': [{'ac...  

[2 rows x 23 columns]

In [161]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2 entries, 0 to 1
Data columns (total 23 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   ID                     2 non-null      object 
 1   DataRow ID             2 non-null      object 
 2   Labeled Data           2 non-null      object 
 3   Label                  2 non-null      object 
 4   Created By             2 non-null      object 
 5   Project Name           2 non-null      object 
 6   Created At             2 non-null      object 
 7   Updated At             2 non-null      object 
 8   Seconds to Label       2 non-null      float64
 9   Seconds to Review      2 non-null      float64
 10  Seconds to Create      2 non-null      float64
 11  External ID            2 non-null      object 
 12  Global Key             0 non-null      object 
 13  Agreement              2 non-null      int64  
 14  Is Benchmark           2 non-null      int64  
 15  Benchmark 

In [162]:
# Check View Label Column
for i in df['View Label']:
    print(i)
  

https://editor.labelbox.com?project=clfbk7q941djz07ymacjs9pqs&label=clfbk8guf04g5071jbrqo26bq
https://editor.labelbox.com?project=clfbk7q941djz07ymacjs9pqs&label=clfbkxgaq0bes07012isv5rq9


In [163]:
# Check Label column
# for i in df['Label']:
    # print(i)

- We need to parse nested dictionaries in df to better understand its structure and to see what keys:values we need to extract;
- The most simple elements are ID, DataRow ID  keys and their values;
- The most complex structure is Label key and its nested value;
    - all annotated data is stored here.

In [164]:
with open('/Users/aidyn/Downloads/export-2023-03-16T21_01_57.705Z.json') as user_file:
  file_contents = user_file.read()
  
# print(file_contents)

parsed_json = json.loads(file_contents)
parsed_json[0]['Label']['objects'][0]

{'featureId': 'clfbk9dbf0002396jfl1k6gpk',
 'schemaId': 'clfbcpdz013ua07zn9k319jn3',
 'color': '#1CE6FF',
 'title': 'uttr',
 'value': 'uttr',
 'version': 1,
 'format': 'text.location',
 'data': {'location': {'start': 151, 'end': 211}},
 'classifications': [{'featureId': 'clfbk9hrj0004396jaxr9fvb7',
   'schemaId': 'clfbcpdz013ub07zn8cw44wpj',
   'title': 'lang',
   'value': 'lang',
   'position': 0,
   'answer': {'featureId': 'clfbk9hrj0003396jlk95obai',
    'schemaId': 'clfbcpdz013uc07zn8vjl7q45',
    'title': 'kz',
    'value': 'kz',
    'position': 0}}]}

In [165]:
type(parsed_json)

list

In [166]:
# Might be helpful to extract keys and values from nested dict json object
def dict_get(x,key,here=None):
    x = x.copy()
    if here is None: here = []
    if x.get(key):  
        here.append(x.get(key))
        x.pop(key)
    else:
        for i,j in x.items():
          if  isinstance(x[i],list): dict_get(x[i][0],key,here)
          if  isinstance(x[i],dict): dict_get(x[i],key,here)
    return here

## Parse nested JSON content

- We want to extract Label key and its nested dictionary values bc annotated text information is stored here;
- The nested JSON file was parsed following this tutorial [here](https://pybit.es/articles/case-study-how-to-parse-nested-json/)

In [167]:
entries = df['Label'][0]['objects']
# entries

In [168]:
entries[0]

{'featureId': 'clfbk9dbf0002396jfl1k6gpk',
 'schemaId': 'clfbcpdz013ua07zn9k319jn3',
 'color': '#1CE6FF',
 'title': 'uttr',
 'value': 'uttr',
 'version': 1,
 'format': 'text.location',
 'data': {'location': {'start': 151, 'end': 211}},
 'classifications': [{'featureId': 'clfbk9hrj0004396jaxr9fvb7',
   'schemaId': 'clfbcpdz013ub07zn8cw44wpj',
   'title': 'lang',
   'value': 'lang',
   'position': 0,
   'answer': {'featureId': 'clfbk9hrj0003396jlk95obai',
    'schemaId': 'clfbcpdz013uc07zn8vjl7q45',
    'title': 'kz',
    'value': 'kz',
    'position': 0}}]}

In [169]:
len(entries)

49

In [170]:
entries2 = df['Label'][1]['objects']
# entries2

In [171]:
label_df1 = pd.DataFrame(entries)
label_df2 = pd.DataFrame(entries2)
label_df = pd.concat([label_df1, label_df2])
label_df.sample(5)

featureId                   schemaId    color title value  \
13  clfbl2hlw0025396jtlcgg1zp  clfbcpdz013ua07zn9k319jn3  #1CE6FF  uttr  uttr   
38  clfbkle630060396j37y42g1k  clfbcpdz013ua07zn9k319jn3  #1CE6FF  uttr  uttr   
20  clfbkgzk1003d396j9cxjz8k3  clfbcpdz013ua07zn9k319jn3  #1CE6FF  uttr  uttr   
6   clfbkcc21001b396jblo0e4fh  clfbcpdz013ua07zn9k319jn3  #1CE6FF  uttr  uttr   
10  clfbkdk1j001v396jynjhh3iu  clfbcpdz013ua07zn9k319jn3  #1CE6FF  uttr  uttr   

    version         format                                        data  \
13        1  text.location  {'location': {'start': 1317, 'end': 1372}}   
38        1  text.location  {'location': {'start': 4314, 'end': 4320}}   
20        1  text.location  {'location': {'start': 2505, 'end': 2573}}   
6         1  text.location    {'location': {'start': 702, 'end': 724}}   
10        1  text.location  {'location': {'start': 1246, 'end': 1270}}   

                                      classifications  
13  [{'featureId': 'clfbl2jl60027396jov9prlm1', 's...  
38  [{'featureId': 'clfbklfbf0062396jgmctg1ed', 's...  
20  [{'featureId': 'clfbkh0zq003f396j0prl16fo', 's...  
6   [{'featureId': 'clfbkcdqq001d396j33e5v6ug', 's...  
10  [{'featureId': 'clfbkdlpp001x396jjasojwnm', 's...

In [172]:
pip install flatten_json

Note: you may need to restart the kernel to use updated packages.


In [173]:
# df_final = pd.DataFrame([flatten_json(data[key]) for key in data])
# df_final

In [174]:
type(label_df)

pandas.core.frame.DataFrame

In [175]:
label_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 111 entries, 0 to 61
Data columns (total 9 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   featureId        111 non-null    object
 1   schemaId         111 non-null    object
 2   color            111 non-null    object
 3   title            111 non-null    object
 4   value            111 non-null    object
 5   version          111 non-null    int64 
 6   format           111 non-null    object
 7   data             111 non-null    object
 8   classifications  111 non-null    object
dtypes: int64(1), object(8)
memory usage: 8.7+ KB


- So,there is 111 annotated samples.
- Classifications column contain nested subtags for each annotated sample, for example:  

    - utterance (uttr) -> language (lang): Kazakh (kz)/Russian (rs)
    - code-switching (cs) -> type: inter-sent/intra-sent/intra-word
        - inter-sent: uttr
        - intra-sent: disc/phr/vp/cl
        - intra-word: n/adj/adv/pn/conj/interj/morph
- More detailed information on the annotation scheme can be found [here](https://github.com/Data-Science-for-Linguists-2023/Kazakh-Russian-Code-Switching-Analysis/blob/main/data-annotated/annotation_scheme_1st_draft.md)
        

In [176]:
dict_get(label_df['classifications'],'title')

['lang',
 'lang',
 'intra-word',
 'intra-word',
 'intra-word',
 'intra-word',
 'intra-word',
 'intra-word',
 'intra-word',
 'intra-word',
 'intra-sent',
 'intra-sent',
 'intra-sent']

In [177]:
len(dict_get(label_df['classifications'],'title'))

13

In [178]:
dict_get(label_df['classifications'],'answer')

[{'featureId': 'clfblh28r008v396julvazgil',
  'schemaId': 'clfbcpdz013uc07zn8vjl7q45',
  'title': 'kz',
  'value': 'kz',
  'position': 0},
 {'featureId': 'clfblh7yh008z396j0c63pt6x',
  'schemaId': 'clfbcpdz013uc07zn8vjl7q45',
  'title': 'kz',
  'value': 'kz',
  'position': 0},
 {'featureId': 'clfbkyhm4000i396j1qjv48vy',
  'schemaId': 'clfbcpdz113uy07zn3wo55wm0',
  'title': 'n',
  'value': 'n',
  'position': 0},
 {'featureId': 'clfbl1nds001r396jp53z541c',
  'schemaId': 'clfbcpdz113v207znepk25gqw',
  'title': 'adv',
  'value': 'adv',
  'position': 2},
 {'featureId': 'clfbl34oh002g396jgg7ra10o',
  'schemaId': 'clfbcpdz113v407znbtfvd85f',
  'title': 'pn',
  'value': 'pn',
  'position': 3},
 {'featureId': 'clfbl4bs30030396j5szz9zr4',
  'schemaId': 'clfbcpdz113v807zn2q2y9ck4',
  'title': 'morph',
  'value': 'morph',
  'position': 5},
 {'featureId': 'clfbl8ccj004t396j7lhkdmid',
  'schemaId': 'clfbcpdz113v807zn2q2y9ck4',
  'title': 'morph',
  'value': 'morph',
  'position': 5},
 {'featureId': 

In [179]:
len(dict_get(label_df['classifications'],'answer'))

13

In [180]:
dict_get(label_df['classifications'],'schemaId')

['clfbcpdz013ub07zn8cw44wpj',
 'clfbcpdz013ub07zn8cw44wpj',
 'clfbcpdz113ux07znfaxvb5r2',
 'clfbcpdz113ux07znfaxvb5r2',
 'clfbcpdz113ux07znfaxvb5r2',
 'clfbcpdz113ux07znfaxvb5r2',
 'clfbcpdz113ux07znfaxvb5r2',
 'clfbcpdz113ux07znfaxvb5r2',
 'clfbcpdz113ux07znfaxvb5r2',
 'clfbcpdz113ux07znfaxvb5r2',
 'clfbcpdz113un07zn0qvugfbt',
 'clfbcpdz113un07zn0qvugfbt',
 'clfbcpdz113un07zn0qvugfbt']

In [181]:
dict_get(label_df['classifications'],'featureId')

['clfblh0n9008u396jtqksj611',
 'clfblh7yh0090396j5b73x4w5',
 'clfbkyhm4000j396jt97cpx6z',
 'clfbl1nds001s396j4bvyqisf',
 'clfbl34oh002h396jmuydz5wv',
 'clfbl4bs30031396jii4jnq80',
 'clfbl8ccj004u396jhv7y644g',
 'clfbl98z50059396jdefdw5wc',
 'clfbla2xm005o396jt8bsgdgv',
 'clfblblsu006i396j44s1550t',
 'clfblc1js006n396jb8lr8kea',
 'clfble333007e396juhak2ekj',
 'clfbleuco007q396j2maybd1e']

In [182]:
len(dict_get(label_df['classifications'],'featureId'))

13

In [183]:
dict_get(label_df['classifications'],'value')

['lang',
 'lang',
 'intra_word',
 'intra_word',
 'intra_word',
 'intra_word',
 'intra_word',
 'intra_word',
 'intra_word',
 'intra_word',
 'intra_sent',
 'intra_sent',
 'intra_sent']

In [184]:
label_df.groupby('schemaId')['title'].value_counts()

schemaId                   title
clfbcpdz013ua07zn9k319jn3  uttr     95
clfbcpdz113ui07znc6r6b37s  cs       16
Name: title, dtype: int64

In [185]:
label_df['title'].value_counts()

uttr    95
cs      16
Name: title, dtype: int64

In [186]:
df = pd.json_normalize(data, sep='.')
display(df)



ID                 DataRow ID  \
0  clfbk8guf04g5071jbrqo26bq  clfbcqidc3xmk07zvgxw373vk   
1  clfbkxgaq0bes07012isv5rq9  clfbkqg468bwf07zc7n31dl3j   

                                        Labeled Data      Created By  \
0  https://storage.labelbox.com/clfagslqv28u407zn...  mob75@pitt.edu   
1  https://storage.labelbox.com/clfagslqv28u407zn...  mob75@pitt.edu   

       Project Name                Created At                Updated At  \
0  annotation-text1  2023-03-16T20:36:16.000Z  2023-03-16T20:36:16.000Z   
1  annotation-text1  2023-03-16T20:59:51.000Z  2023-03-16T20:59:51.000Z   

   Seconds to Label  Seconds to Review  Seconds to Create  ...  \
0           691.644             32.957            658.687  ...   
1           926.531              0.000            926.531  ...   

     Dataset Name Reviews                                         View Label  \
0  cs-annotation1      []  https://editor.labelbox.com?project=clfbk7q941...   
1           text2      []  https://editor.labelbox.com?project=clfbk7q941...   

   Has Open Issues  Skipped  \
0                0    False   
1                0    False   

                                       Label.objects Label.classifications  \
0  [{'featureId': 'clfbk9dbf0002396jfl1k6gpk', 's...                    []   
1  [{'featureId': 'clfbkxma70002396jwrrwp26n', 's...                    []   

  Label.relationships DataRow Workflow Info.taskName  \
0                  []                           Done   
1                  []                           Done   

              DataRow Workflow Info.Workflow History  
0  [{'actorId': 'clfagslum28u607zn20th1y45', 'act...  
1  [{'actorId': 'clfagslum28u607zn20th1y45', 'act...  

[2 rows x 26 columns]

In [187]:
# Concat two dict-entries
entries = entries + entries2
entries[0]

{'featureId': 'clfbk9dbf0002396jfl1k6gpk',
 'schemaId': 'clfbcpdz013ua07zn9k319jn3',
 'color': '#1CE6FF',
 'title': 'uttr',
 'value': 'uttr',
 'version': 1,
 'format': 'text.location',
 'data': {'location': {'start': 151, 'end': 211}},
 'classifications': [{'featureId': 'clfbk9hrj0004396jaxr9fvb7',
   'schemaId': 'clfbcpdz013ub07zn8cw44wpj',
   'title': 'lang',
   'value': 'lang',
   'position': 0,
   'answer': {'featureId': 'clfbk9hrj0003396jlk95obai',
    'schemaId': 'clfbcpdz013uc07zn8vjl7q45',
    'title': 'kz',
    'value': 'kz',
    'position': 0}}]}

In [188]:
type(entries)

list

In [189]:
# parsed_data = defaultdict(list)

def recursive_parser(entry: dict, data_dict: dict, col_name: str = "") -> dict:
    """Recursive parser for a list of nested JSON objects
    
    Args:
        entry (dict): A dictionary representing a single entry (row) of the final data frame.
        data_dict (dict): Accumulator holding the current parsed data.
        col_name (str): Accumulator holding the current column name. Defaults to empty string.
    """
    for key, val in entry.items():
        extended_col_name = f"{col_name}_{key}" if col_name else key
        if isinstance(val, dict):
            recursive_parser(entry[key], data_dict, extended_col_name)
        else:
            data_dict[extended_col_name].append(val)

parsed_data = defaultdict(list)

for entry in entries:
    recursive_parser(entry, parsed_data, "")

df = pd.DataFrame(parsed_data)

In [190]:
df

featureId                   schemaId    color title  \
0    clfbk9dbf0002396jfl1k6gpk  clfbcpdz013ua07zn9k319jn3  #1CE6FF  uttr   
1    clfbk9q660007396ji9fkn5ya  clfbcpdz013ua07zn9k319jn3  #1CE6FF  uttr   
2    clfbkajp7000h396jticb6tb7  clfbcpdz013ua07zn9k319jn3  #1CE6FF  uttr   
3    clfbkaq3k000m396jjwbnfb8n  clfbcpdz013ua07zn9k319jn3  #1CE6FF  uttr   
4    clfbkb19x000r396jmixhcf23  clfbcpdz013ua07zn9k319jn3  #1CE6FF  uttr   
..                         ...                        ...      ...   ...   
106  clfbla0ed005m396jhwxzv2ui  clfbcpdz113ui07znc6r6b37s  #FF34FF    cs   
107  clfblbjlu006g396jgex1r7gu  clfbcpdz113ui07znc6r6b37s  #FF34FF    cs   
108  clfblbx1q006l396jc40wocvd  clfbcpdz113ui07znc6r6b37s  #FF34FF    cs   
109  clfble0yu007c396ja0be4j2m  clfbcpdz113ui07znc6r6b37s  #FF34FF    cs   
110  clfblepxv007o396jjnd6u8d0  clfbcpdz113ui07znc6r6b37s  #FF34FF    cs   

    value  version         format  data_location_start  data_location_end  \
0    uttr        1  text.location                  151                211   
1    uttr        1  text.location                  228                296   
2    uttr        1  text.location                  313                387   
3    uttr        1  text.location                  398                413   
4    uttr        1  text.location                  429                458   
..    ...      ...            ...                  ...                ...   
106    cs        1  text.location                 4529               4541   
107    cs        1  text.location                 4976               4981   
108    cs        1  text.location                 5033               5049   
109    cs        1  text.location                 5620               5624   
110    cs        1  text.location                 5643               5646   

                                       classifications  
0    [{'featureId': 'clfbk9hrj0004396jaxr9fvb7', 's...  
1    [{'featureId': 'clfbk9sie0009396jvshpmff7', 's...  
2    [{'featureId': 'clfbkal14000j396jo0ry3eb8', 's...  
3    [{'featureId': 'clfbkarhn000o396j0lao0sgk', 's...  
4    [{'featureId': 'clfbkb2up000t396jvhwtdk5a', 's...  
..                                                 ...  
106  [{'featureId': 'clfbla2xm005o396jt8bsgdgv', 's...  
107  [{'featureId': 'clfblblsu006i396j44s1550t', 's...  
108  [{'featureId': 'clfblc1js006n396jb8lr8kea', 's...  
109  [{'featureId': 'clfble333007e396juhak2ekj', 's...  
110  [{'featureId': 'clfbleuco007q396j2maybd1e', 's...  

[111 rows x 10 columns]

In [191]:
for i in df['classifications'][0]:
    print(i)

{'featureId': 'clfbk9hrj0004396jaxr9fvb7', 'schemaId': 'clfbcpdz013ub07zn8cw44wpj', 'title': 'lang', 'value': 'lang', 'position': 0, 'answer': {'featureId': 'clfbk9hrj0003396jlk95obai', 'schemaId': 'clfbcpdz013uc07zn8vjl7q45', 'title': 'kz', 'value': 'kz', 'position': 0}}


In [192]:
for i in df['classifications'][110]:
    print(i)

{'featureId': 'clfbleuco007q396j2maybd1e', 'schemaId': 'clfbcpdz113un07zn0qvugfbt', 'title': 'intra-sent', 'value': 'intra_sent', 'position': 1, 'answer': {'featureId': 'clfbleuco007p396jhonhjzf6', 'schemaId': 'clfbcpdz113uo07znh96wdg4o', 'title': 'disc', 'value': 'disc', 'position': 0}}


In [193]:
dic = df['classifications'].to_dict()
dic[0]

[{'featureId': 'clfbk9hrj0004396jaxr9fvb7',
  'schemaId': 'clfbcpdz013ub07zn8cw44wpj',
  'title': 'lang',
  'value': 'lang',
  'position': 0,
  'answer': {'featureId': 'clfbk9hrj0003396jlk95obai',
   'schemaId': 'clfbcpdz013uc07zn8vjl7q45',
   'title': 'kz',
   'value': 'kz',
   'position': 0}}]

In [194]:
type(dic)

dict

In [195]:
flattened_clf = flatten(dic)

In [196]:
import flatten_json
def flatten_json(nested_json: dict, exclude: list=[''], sep: str='_') -> dict:
    """
    Flatten a list of nested dicts.
    """
    out = dict()
    def flatten(x: (list, dict, str), name: str='', exclude=exclude):
        if type(x) is dict:
            for a in x:
                if a not in exclude:
                    flatten(x[a], f'{name}{a}{sep}')
        elif type(x) is list:
            i = 0
            for a in x:
                flatten(a, f'{name}{i}{sep}')
                i += 1
        else:
            out[name[:-1]] = x

    flatten(nested_json)
    return out

flatten_labeldf = flatten_json(label_df)

In [197]:
type(flatten_labeldf)

dict

In [198]:
flatten_labeldf.items()

dict_items([('',                     featureId                   schemaId    color title value  \
0   clfbk9dbf0002396jfl1k6gpk  clfbcpdz013ua07zn9k319jn3  #1CE6FF  uttr  uttr   
1   clfbk9q660007396ji9fkn5ya  clfbcpdz013ua07zn9k319jn3  #1CE6FF  uttr  uttr   
2   clfbkajp7000h396jticb6tb7  clfbcpdz013ua07zn9k319jn3  #1CE6FF  uttr  uttr   
3   clfbkaq3k000m396jjwbnfb8n  clfbcpdz013ua07zn9k319jn3  #1CE6FF  uttr  uttr   
4   clfbkb19x000r396jmixhcf23  clfbcpdz013ua07zn9k319jn3  #1CE6FF  uttr  uttr   
..                        ...                        ...      ...   ...   ...   
57  clfbla0ed005m396jhwxzv2ui  clfbcpdz113ui07znc6r6b37s  #FF34FF    cs    cs   
58  clfblbjlu006g396jgex1r7gu  clfbcpdz113ui07znc6r6b37s  #FF34FF    cs    cs   
59  clfblbx1q006l396jc40wocvd  clfbcpdz113ui07znc6r6b37s  #FF34FF    cs    cs   
60  clfble0yu007c396ja0be4j2m  clfbcpdz113ui07znc6r6b37s  #FF34FF    cs    cs   
61  clfblepxv007o396jjnd6u8d0  clfbcpdz113ui07znc6r6b37s  #FF34FF    cs    cs   

    versio

In [259]:
# entry_dict = flatten_labeldf['classifications'][0]['answer']
# entry_dict
# flatten_labeldf.keys()

dict_keys([''])

#### Setting global tags

In [199]:
# reading JSON file
df_pilot5 = pd.read_json('/Users/aidyn/Downloads/export-2023-03-18T18_44_16.835Z.json')

# displaying sample output
df_pilot5.head()

ID                 DataRow ID  \
0  clfeba39t08r707y49iyjb6rd  clfbkqg468bwf07zc7n31dl3j   

                                        Labeled Data  \
0  https://storage.labelbox.com/clfagslqv28u407zn...   

                                               Label      Created By  \
0  {'objects': [{'featureId': 'clfebao350002396jb...  mob75@pitt.edu   

         Project Name                Created At                Updated At  \
0  annotation-pilot-5  2023-03-18T18:43:26.000Z  2023-03-18T18:43:26.000Z   

   Seconds to Label  Seconds to Review  ...  Agreement Is Benchmark  \
0           333.261                  0  ...         -1            0   

   Benchmark Agreement  Benchmark ID  Dataset Name  Reviews  \
0                   -1           NaN         text2       []   

                                          View Label Has Open Issues Skipped  \
0  https://editor.labelbox.com?project=clfeazj2x0...               0   False   

                               DataRow Workflow Info  
0  {'taskId': 'db06a4f6-ad4d-4ff0-b90e-4d0c0a701b...  

[1 rows x 23 columns]

In [200]:
labels = df_pilot5['Label'][0]['objects']
labels

[{'featureId': 'clfebao350002396jb23w7a83',
  'schemaId': 'clfeayddt0q3e07346y4fcus3',
  'color': '#1CE6FF',
  'title': 'uttr',
  'value': 'uttr',
  'version': 1,
  'format': 'text.location',
  'data': {'location': {'start': 26, 'end': 42}}},
 {'featureId': 'clfebdyk2000d396jx2mihrxi',
  'schemaId': 'clfeayddt0q3e07346y4fcus3',
  'color': '#1CE6FF',
  'title': 'uttr',
  'value': 'uttr',
  'version': 1,
  'format': 'text.location',
  'data': {'location': {'start': 106, 'end': 124}}},
 {'featureId': 'clfebeb8m000g396j2vg1hddw',
  'schemaId': 'clfeayddt0q3e07346y4fcus3',
  'color': '#1CE6FF',
  'title': 'uttr',
  'value': 'uttr',
  'version': 1,
  'format': 'text.location',
  'data': {'location': {'start': 168, 'end': 259}}},
 {'featureId': 'clfebfl84000n396jcbjyjln3',
  'schemaId': 'clfeayddt0q3e07346y4fcus3',
  'color': '#1CE6FF',
  'title': 'uttr',
  'value': 'uttr',
  'version': 1,
  'format': 'text.location',
  'data': {'location': {'start': 423, 'end': 437}}},
 {'featureId': 'clfebg

In [201]:
pilot_df = pd.DataFrame(labels)
pilot_df

featureId                   schemaId    color title value  \
0  clfebao350002396jb23w7a83  clfeayddt0q3e07346y4fcus3  #1CE6FF  uttr  uttr   
1  clfebdyk2000d396jx2mihrxi  clfeayddt0q3e07346y4fcus3  #1CE6FF  uttr  uttr   
2  clfebeb8m000g396j2vg1hddw  clfeayddt0q3e07346y4fcus3  #1CE6FF  uttr  uttr   
3  clfebfl84000n396jcbjyjln3  clfeayddt0q3e07346y4fcus3  #1CE6FF  uttr  uttr   
4  clfebg02r000r396jdkxwm5e5  clfeayddt0q3e07346y4fcus3  #1CE6FF  uttr  uttr   
5  clfebga26000s396j3lrycm8n  clfeayddt0q3e07346y4fcus3  #1CE6FF  uttr  uttr   
6  clfebghe7000t396ji7mharwj  clfeayddt0q3e07346y4fcus3  #1CE6FF  uttr  uttr   
7  clfebgoo2000u396jjdtn9h4c  clfeayddt0q3e07346y4fcus3  #1CE6FF  uttr  uttr   
8  clfebgv7m000v396jjnr7q3am  clfeayddt0q3e07346y4fcus3  #1CE6FF  uttr  uttr   

   version         format                                       data  
0        1  text.location     {'location': {'start': 26, 'end': 42}}  
1        1  text.location   {'location': {'start': 106, 'end': 124}}  
2        1  text.location   {'location': {'start': 168, 'end': 259}}  
3        1  text.location   {'location': {'start': 423, 'end': 437}}  
4        1  text.location   {'location': {'start': 561, 'end': 585}}  
5        1  text.location   {'location': {'start': 679, 'end': 705}}  
6        1  text.location   {'location': {'start': 793, 'end': 821}}  
7        1  text.location   {'location': {'start': 874, 'end': 902}}  
8        1  text.location  {'location': {'start': 934, 'end': 1008}}

- Note: using global tags overrides the code-switching sub-classifications.

## Test sample 2

In [203]:
# reading JSON file
df = pd.read_json('/Users/aidyn/Documents/Data_Science/Kazakh-Russian-Code-Switching-Analysis/annotated-data-samples/export-2023-03-19T20_19_44.671Z.json')

# displaying sample output
df.head()


ID                 DataRow ID  \
0  clfbk8guf04g5071jbrqo26bq  clfbcqidc3xmk07zvgxw373vk   
1  clfbkxgaq0bes07012isv5rq9  clfbkqg468bwf07zc7n31dl3j   
2  clfef6xx80t7207yzh33ohlgx  clfeb1kdn1yhb07895fdzagon   
3  clfeh5eam0l8g07zr7djaedpq  clfeb7o4k17e7079p5eo5hd9f   
4  clffjz71f17oh07yxdfyo65zo  clffjq06201n907a87jni6dom   

                                        Labeled Data  \
0  https://storage.labelbox.com/clfagslqv28u407zn...   
1  https://storage.labelbox.com/clfagslqv28u407zn...   
2  https://storage.labelbox.com/clfagslqv28u407zn...   
3  https://storage.labelbox.com/clfagslqv28u407zn...   
4  https://storage.labelbox.com/clfagslqv28u407zn...   

                                               Label      Created By  \
0  {'objects': [{'featureId': 'clfbk9dbf0002396jf...  mob75@pitt.edu   
1  {'objects': [{'featureId': 'clfbkxma70002396jw...  mob75@pitt.edu   
2  {'objects': [{'featureId': 'clfef75jw0002396jb...  mob75@pitt.edu   
3  {'objects': [{'featureId': 'clfeh5naw0002396jf...  mob75@pitt.edu   
4  {'objects': [{'featureId': 'clffk6vxv0002396jh...  mob75@pitt.edu   

            Project Name                Created At                Updated At  \
0  cs-annotation-project  2023-03-16T20:36:16.000Z  2023-03-16T20:36:16.000Z   
1  cs-annotation-project  2023-03-16T20:59:51.000Z  2023-03-16T20:59:51.000Z   
2  cs-annotation-project  2023-03-18T20:54:54.000Z  2023-03-18T20:54:54.000Z   
3  cs-annotation-project  2023-03-18T21:33:53.000Z  2023-03-19T14:32:23.000Z   
4  cs-annotation-project  2023-03-19T15:48:36.000Z  2023-03-19T15:48:36.000Z   

   Seconds to Label  Seconds to Review  ...  Agreement Is Benchmark  \
0           749.159             72.967  ...         -1            0   
1           972.239              0.000  ...         -1            0   
2          1664.856              0.000  ...         -1            0   
3          1196.039              0.000  ...         -1            0   
4           840.414              0.000  ...         -1            0   

   Benchmark Agreement  Benchmark ID    Dataset Name  Reviews  \
0                   -1           NaN  cs-annotation1       []   
1                   -1           NaN           text2       []   
2                   -1           NaN           text5       []   
3                   -1           NaN           text6       []   
4                   -1           NaN          text11       []   

                                          View Label Has Open Issues Skipped  \
0  https://editor.labelbox.com?project=clfbk7q941...               0   False   
1  https://editor.labelbox.com?project=clfbk7q941...               0   False   
2  https://editor.labelbox.com?project=clfbk7q941...               0   False   
3  https://editor.labelbox.com?project=clfbk7q941...               0   False   
4  https://editor.labelbox.com?project=clfbk7q941...               0   False   

                               DataRow Workflow Info  
0  {'taskName': 'Done', 'Workflow History': [{'ac...  
1  {'taskName': 'Done', 'Workflow History': [{'ac...  
2  {'taskId': '84a4bf5c-a482-464a-8623-da3757c0f5...  
3  {'taskName': 'Done', 'Workflow History': [{'ac...  
4  {'taskName': 'Done', 'Workflow History': [{'ac...  

[5 rows x 23 columns]

In [204]:
for i in df['Labeled Data'][:3]:
    print(i)

https://storage.labelbox.com/clfagslqv28u407zn55sthytz%2F4669a93f-90a9-e7bc-869e-6674927e3870-BABEL_OP2_302_96842_20140131_154710_inLine.txt?Expires=1680466786135&KeyName=labelbox-assets-key-3&Signature=u3gipVbP58L93HUfcgCFcd3EtqU
https://storage.labelbox.com/clfagslqv28u407zn55sthytz%2Fb1f1e43f-d07a-f08e-8ba1-317c48254edb-BABEL_OP2_302_95583_20131112_203137_outLine.txt?Expires=1680466786139&KeyName=labelbox-assets-key-3&Signature=kU7HEEcR6jNPul5lv48-Fg4Tl6k
https://storage.labelbox.com/clfagslqv28u407zn55sthytz%2F140ef9d3-6f66-97aa-86fe-7c1064aefd68-BABEL_OP2_302_93475_20131115_203137_outLine.txt?Expires=1680466786140&KeyName=labelbox-assets-key-3&Signature=k-eN8XWROzIZ7VMxtXvXunIwAYU


- Cyrillic characters are not recognized!

In [205]:
for i in df['View Label'][:3]:
    print(i)

https://editor.labelbox.com?project=clfbk7q941djz07ymacjs9pqs&label=clfbk8guf04g5071jbrqo26bq
https://editor.labelbox.com?project=clfbk7q941djz07ymacjs9pqs&label=clfbkxgaq0bes07012isv5rq9
https://editor.labelbox.com?project=clfbk7q941djz07ymacjs9pqs&label=clfef6xx80t7207yzh33ohlgx


- I can view each annotated file and show examples..

In [206]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14 entries, 0 to 13
Data columns (total 23 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   ID                     14 non-null     object 
 1   DataRow ID             14 non-null     object 
 2   Labeled Data           14 non-null     object 
 3   Label                  14 non-null     object 
 4   Created By             14 non-null     object 
 5   Project Name           14 non-null     object 
 6   Created At             14 non-null     object 
 7   Updated At             14 non-null     object 
 8   Seconds to Label       14 non-null     float64
 9   Seconds to Review      14 non-null     float64
 10  Seconds to Create      14 non-null     float64
 11  External ID            14 non-null     object 
 12  Global Key             0 non-null      float64
 13  Agreement              14 non-null     int64  
 14  Is Benchmark           14 non-null     int64  
 15  Benchmar

In [207]:
# Extract only necessary columns
df_test2_old = df[['Label', 'Seconds to Label', 'External ID','View Label']]
df_test2 = df_test2_old[['External ID','View Label', 'Seconds to Label', 'Label']].rename(columns={'View Label':'View Tag', 'Seconds to Label':'Seconds to Tag','Label':'Tags'}) 
df_test2.sample(5)

External ID  \
3   BABEL_OP2_302_93320_20140218_173001_outLine.txt   
10   BABEL_OP2_302_86557_20131121_000022_inLine.txt   
2   BABEL_OP2_302_93475_20131115_203137_outLine.txt   
13  BABEL_OP2_302_87889_20140119_163150_outLine.txt   
1   BABEL_OP2_302_95583_20131112_203137_outLine.txt   

                                             View Tag  Seconds to Tag  \
3   https://editor.labelbox.com?project=clfbk7q941...        1196.039   
10  https://editor.labelbox.com?project=clfbk7q941...         965.589   
2   https://editor.labelbox.com?project=clfbk7q941...        1664.856   
13  https://editor.labelbox.com?project=clfbk7q941...         737.941   
1   https://editor.labelbox.com?project=clfbk7q941...         972.239   

                                                 Tags  
3   {'objects': [{'featureId': 'clfeh5naw0002396jf...  
10  {'objects': [{'featureId': 'clffrealm0002396jg...  
2   {'objects': [{'featureId': 'clfef75jw0002396jb...  
13  {'objects': [{'featureId': 'clfftuhqt012g396j3...  
1   {'objects': [{'featureId': 'clfbkxma70002396jw...

In [208]:
df_test2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14 entries, 0 to 13
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   External ID     14 non-null     object 
 1   View Tag        14 non-null     object 
 2   Seconds to Tag  14 non-null     float64
 3   Tags            14 non-null     object 
dtypes: float64(1), object(3)
memory usage: 576.0+ bytes


In [286]:
df_test2['Tagged Rows'] = pd.Series(df_test2['Tags'][0]['objects']).map(lambda x:len(x))
df_test2.sample(5)
# len(df_test2['Tags'][0]['objects'])

External ID  \
3   BABEL_OP2_302_93320_20140218_173001_outLine.txt   
9   BABEL_OP2_302_90080_20140120_230635_outLine.txt   
8    BABEL_OP2_302_90080_20140120_230635_inLine.txt   
6   BABEL_OP2_302_91593_20140215_175049_outLine.txt   
11  BABEL_OP2_302_86557_20131121_000022_outLine.txt   

                                             View Tag  Seconds to Tag  \
3   https://editor.labelbox.com?project=clfbk7q941...        1196.039   
9   https://editor.labelbox.com?project=clfbk7q941...         535.854   
8   https://editor.labelbox.com?project=clfbk7q941...         698.394   
6   https://editor.labelbox.com?project=clfbk7q941...         830.769   
11  https://editor.labelbox.com?project=clfbk7q941...        1615.877   

                                                 Tags  Tagged Rows  
3   {'objects': [{'featureId': 'clfeh5naw0002396jf...            9  
9   {'objects': [{'featureId': 'clffqq2zj00bd396jx...            9  
8   {'objects': [{'featureId': 'clffqb8la0002396jl...            9  
6   {'objects': [{'featureId': 'clffmg0sy0002396jx...            9  
11  {'objects': [{'featureId': 'clffrwdo900ai396jf...            9

In [238]:
df_test2['Tagged Rows'][0]

{'featureId': 'clfbk9dbf0002396jfl1k6gpk',
 'schemaId': 'clfbcpdz013ua07zn9k319jn3',
 'color': '#1CE6FF',
 'title': 'uttr',
 'value': 'uttr',
 'version': 1,
 'format': 'text.location',
 'data': {'location': {'start': 151, 'end': 211}},
 'classifications': [{'featureId': 'clfbk9hrj0004396jaxr9fvb7',
   'schemaId': 'clfbcpdz013ub07zn8cw44wpj',
   'title': 'lang',
   'value': 'lang',
   'position': 0,
   'answer': {'featureId': 'clfbk9hrj0003396jlk95obai',
    'schemaId': 'clfbcpdz013uc07zn8vjl7q45',
    'title': 'kz',
    'value': 'kz',
    'position': 0}}]}

In [236]:
type(pd.Series(df_test2['Tags'][0]['objects']))

pandas.core.series.Series

In [248]:
df_test2['Tags'][0]['objects'][0].items()

dict_items([('featureId', 'clfbk9dbf0002396jfl1k6gpk'), ('schemaId', 'clfbcpdz013ua07zn9k319jn3'), ('color', '#1CE6FF'), ('title', 'uttr'), ('value', 'uttr'), ('version', 1), ('format', 'text.location'), ('data', {'location': {'start': 151, 'end': 211}}), ('classifications', [{'featureId': 'clfbk9hrj0004396jaxr9fvb7', 'schemaId': 'clfbcpdz013ub07zn8cw44wpj', 'title': 'lang', 'value': 'lang', 'position': 0, 'answer': {'featureId': 'clfbk9hrj0003396jlk95obai', 'schemaId': 'clfbcpdz013uc07zn8vjl7q45', 'title': 'kz', 'value': 'kz', 'position': 0}}])])

In [241]:
df_test2['Tags'][0]['objects'][0]['featureId']

'clfbk9dbf0002396jfl1k6gpk'

In [243]:
len(df_test2['Tags'][0])

3

### Parse Tags column

In [209]:
for i in df_test2['Tags'][0]:
    print(i)

objects
classifications
relationships


In [210]:
tags_df_1 = df_test2['Tags'][0]['objects']
tags_df_2 = df_test2['Tags'][1]['objects']
tags_df_3 = df_test2['Tags'][2]['objects']
tags_df_4 = df_test2['Tags'][3]['objects']
tags_df_5 = df_test2['Tags'][4]['objects']
tags_df_6 = df_test2['Tags'][5]['objects']
tags_df_7 = df_test2['Tags'][6]['objects']
tags_df_8 = df_test2['Tags'][7]['objects']
tags_df_9 = df_test2['Tags'][8]['objects']
tags_df_10 = df_test2['Tags'][9]['objects']
tags_df_11 = df_test2['Tags'][10]['objects']
tags_df_12 = df_test2['Tags'][11]['objects']
tags_df_13 = df_test2['Tags'][12]['objects']
tags_df_14 = df_test2['Tags'][13]['objects']

tags_list = tags_df_1 + tags_df_2 + tags_df_3 + tags_df_4 + tags_df_5 + tags_df_6 + tags_df_7 + tags_df_8 + tags_df_9 + tags_df_10  + tags_df_11 + tags_df_12 + tags_df_13 + tags_df_14

tags_df = pd.DataFrame(tags_list)
tags_df.sample(5)

featureId                   schemaId    color title  \
493  clffmub6l0094396jg2c07qov  clfbcpdz013ua07zn9k319jn3  #1CE6FF  uttr   
2    clfbkajp7000h396jticb6tb7  clfbcpdz013ua07zn9k319jn3  #1CE6FF  uttr   
703  clffrezj2000m396jki7k0cft  clfbcpdz013ua07zn9k319jn3  #1CE6FF  uttr   
618  clffqmot90090396jmaha6mcb  clfbcpdz013ua07zn9k319jn3  #1CE6FF  uttr   
725  clffrl59i004j396j0yfysa3y  clfbcpdz013ua07zn9k319jn3  #1CE6FF  uttr   

    value  version         format                                        data  \
493  uttr        1  text.location    {'location': {'start': 171, 'end': 188}}   
2    uttr        1  text.location    {'location': {'start': 313, 'end': 387}}   
703  uttr        1  text.location    {'location': {'start': 341, 'end': 352}}   
618  uttr        1  text.location  {'location': {'start': 5204, 'end': 5229}}   
725  uttr        1  text.location  {'location': {'start': 2652, 'end': 2702}}   

                                       classifications  
493  [{'featureId': 'clffmucj90096396jvagktll9', 's...  
2    [{'featureId': 'clfbkal14000j396jo0ry3eb8', 's...  
703  [{'featureId': 'clffrf12k000o396ja06rjtk8', 's...  
618  [{'featureId': 'clffqmpwi0092396jmy4o2ktv', 's...  
725  [{'featureId': 'clffrl6pz004l396jnph3bv4u', 's...

In [211]:
len(tags_df)

1052

In [212]:
tags_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1052 entries, 0 to 1051
Data columns (total 9 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   featureId        1052 non-null   object
 1   schemaId         1052 non-null   object
 2   color            1052 non-null   object
 3   title            1052 non-null   object
 4   value            1052 non-null   object
 5   version          1052 non-null   int64 
 6   format           1052 non-null   object
 7   data             1052 non-null   object
 8   classifications  1051 non-null   object
dtypes: int64(1), object(8)
memory usage: 74.1+ KB


In [213]:
tag_dic = df_test2['Tags'].to_dict()
len(tag_dic[0])

3

In [214]:
len(tag_dic)

14

In [215]:
# tag_dic['Label']['objects']['classifications']

In [216]:
tags_df['title'].value_counts()

uttr    905
cs      147
Name: title, dtype: int64

In [217]:
answer_list = dict_get(tags_df['classifications'],'answer')

In [218]:
answer_df = pd.DataFrame(answer_list)
answer_df 

featureId                   schemaId  title  value  \
0     clfbk9hrj0003396jlk95obai  clfbcpdz013uc07zn8vjl7q45     kz     kz   
1     clfbk9sie0008396jckkidvdl  clfbcpdz013uc07zn8vjl7q45     kz     kz   
2     clfbkal14000i396jlhysk84v  clfbcpdz013uc07zn8vjl7q45     kz     kz   
3     clfbkarhn000n396j20edd2if  clfbcpdz013uc07zn8vjl7q45     kz     kz   
4     clfbkb2up000s396j4qiaybx0  clfbcpdz013uc07zn8vjl7q45     kz     kz   
...                         ...                        ...    ...    ...   
1046  clffu9eh401cw396jmex3ex42  clfbcpdz013uc07zn8vjl7q45     kz     kz   
1047  clffu9lz101d1396j5y4ygjft  clfbcpdz013uc07zn8vjl7q45     kz     kz   
1048  clffu9uu201d6396j4f8zy178  clfbcpdz013uc07zn8vjl7q45     kz     kz   
1049  clffu744c01bn396jtuxtxlqp  clfbcpdz113v807zn2q2y9ck4  morph  morph   
1050  clffu96e101cr396jzm1m8a78  clfbcpdz113uo07znh96wdg4o   disc   disc   

      position  
0            0  
1            0  
2            0  
3            0  
4            0  
...        ...  
1046         0  
1047         0  
1048         0  
1049         5  
1050         0  

[1051 rows x 5 columns]

In [219]:
answer_df['title'].value_counts()

kz        896
morph      43
adv        35
n          19
interj     13
disc       11
rs          9
phr         8
vp          5
adj         5
conj        4
pn          2
uttr        1
Name: title, dtype: int64

In [220]:
title_list = dict_get(tags_df['classifications'],'title')

In [221]:
len(title_list)

1051

In [222]:
title_df = pd.DataFrame(title_list)
title_df.value_counts()

lang          905
intra-word    121
intra-sent     24
inter-sent      1
dtype: int64

In [223]:
# title_df

In [224]:
value_list = dict_get(tags_df['classifications'],'value')
value_df = pd.DataFrame(value_list)
value_df.value_counts()

lang          905
intra_word    121
intra_sent     24
inter_sent      1
dtype: int64